volatility3/framework/symbols/linux/__init__.py

```diff
@@ -59,83 +59,92 @@ class LinuxUtilities(interfaces.configuration.VersionableInterface):
+        rdentry = fs_root.dentry
+        rmnt = fs_root.mnt
+        vfsmnt = mnt.mnt
+        dentry = vfsmnt.mnt_root
 
-                dentry = vfsmnt.get_mnt_mountpoint()
-                vfsmnt = vfsmnt.get_mnt_parent()
+        return cls.do_get_path(rdentry, rmnt, dentry, vfsmnt, context)
 
-                continue
+    @classmethod
+    def _get_vmlinux_from_volobj(cls, volobj, context):
+        symbol_table_arr = volobj.vol.type_name.split("!", 1)
+        symbol_table = symbol_table_arr[0] if len(symbol_table_arr) == 2 else None
 
-            parent = dentry.d_parent
-            dentry = parent
+        module_names = context.modules.get_modules_by_symbol_tables(symbol_table)
+        module_names = list(module_names)
 
-        # if we did not gather any valid dentrys in the path, then the entire file is
-        # either 1) smeared out of memory or 2) de-allocated and corresponding structures overwritten
-        # we return an empty string in this case to avoid confusion with something like a handle to the root
-        # directory (e.g., "/")
-        if not ret_path:
-            return ""
+        if not module_names:
+            raise ValueError(f"No module using the symbol table '{symbol_table}'")
 
-        ret_val = "/".join([str(p) for p in ret_path if p != ""])
+        kernel_module_name = module_names[0]
+        kernel = context.modules[kernel_module_name]
 
-        if ret_val.startswith(("socket:", "pipe:")):
-            if ret_val.find("]") == -1:
-                try:
-                    inode = dentry.d_inode
-                    ino = inode.i_ino
-                except exceptions.InvalidAddressException:
-                    ino = 0
+        return kernel
 
-                ret_val = ret_val[:-1] + f":[{ino}]"
-            else:
-                ret_val = ret_val.replace("/", "")
+    @classmethod
+    def _get_mnt_from_vfsmnt(cls, vfsmnt, dentry, context):
+        vmlinux = cls._get_vmlinux_from_volobj(dentry, context)
+
+        # When it's called from _get_path_file(), 'vfsmnt' is a Pointer
+        # struct file->f_path->mnt is "struct vfsmount *".
+        # However, when called from _get_path_root()
+        # struct mount -> mnt is "struct vfsmount"
+        vfsmnt_ptr = vfsmnt if type(vfsmnt) == Pointer else vfsmnt.vol.offset
 
-        elif ret_val != "inotify":
-            ret_val = "/" + ret_val
+        mnt = cls.container_of(vfsmnt_ptr, "mount", "mnt", vmlinux)
 
-        return ret_val
+        return mnt
 
-    # method used by 'older' kernels
-    # TODO: lookup when dentry_operations->d_name was merged into the mainline kernel for exact version
     @classmethod
-    def _get_path_file(cls, task, filp) -> str:
-        rdentry = task.fs.get_root_dentry()
-        rmnt = task.fs.get_root_mnt()
-        dentry = filp.get_dentry()
-        vfsmnt = filp.get_vfsmnt()
+    def do_get_path(cls, rdentry, rmnt, dentry, vfsmnt, context) -> Union[None, str]:
```

Comment:
Also, our convention has typically been to put context as the very first parameter to a method like this?  Since we don't have any backwards compatibility and we're bumping the version number anyway, it'd be good to fix it now so we don't scratch our heads as to how it came to be down the line please...  5:)